lib/puppet/http/service/file_server.rb

```diff
@@ -0,0 +1,103 @@
+require 'puppet/file_serving/metadata'
+
+class Puppet::HTTP::Service::FileServer < Puppet::HTTP::Service
+  API = '/puppet/v3'.freeze
+
+  EXCLUDED_FORMATS = [:yaml, :b64_zlib_yamli, :dot]
+
+  def initialize(client, server, port)
+    url = build_url(API, server || Puppet[:server], port || Puppet[:masterport])
+    super(client, url)
+  end
+
+  def get_file_metadata(mount_point:, path:, environment:,
+                        links: :manage, checksum_type: Puppet[:digest_algorithm], source_permissions: :ignore,
+                        ssl_context: nil)
+
+    model = Puppet::Transaction::Report
+    network_formats = model.supported_formats - EXCLUDED_FORMATS
+    mime_types = network_formats.map { |f| model.get_format(f).mime }
+
+    response = @client.get(
+      with_base_url("/file_metadata/#{mount_point}/#{path}"),
+      headers: add_puppet_headers({ 'ACCEPT' => mime_types.join(', ') }),
+      params: {
+        :links => links,
+        :checksum_type => checksum_type,
+        :source_permissions => source_permissions,
+        :environment => environment
+      },
+      ssl_context: ssl_context
+    )
+
+
+    if response.success?
+      header = response['Content-Type']
+      if header
+        header.gsub!(/\s*;.*$/,'') # strip any charset
+        formatter = Puppet::Network::FormatHandler.mime(header)
+        raise Puppet::HTTP::ProtocolError.new("Content-Type is unsupported") if EXCLUDED_FORMATS.include?(formatter.name)
+      else
+        raise Puppet::HTTP::ProtocolError.new(_("No content type in http response; cannot parse"))
+      end
+      Puppet::FileServing::Metadata.convert_from(formatter.name, response.body.to_s)
+    else
+      raise Puppet::HTTP::ResponseError.new(response)
+    end
```

Comment:
Recommend moving this into a private method, which returns a formatter for the response, then this could be simplified to:

```ruby
raise Puppet::HTTP::ResponseError.new(response) unless response.success?

formatter = formatter_for_response(response)
Puppet::FileServing::Metadata.convert_from(formatter.name, response.body.to_s)
```

and all of the content-type error handling is in the `formatter_for_response` method.